classes/ActionScheduler_Logger.php

```diff
@@ -41,7 +41,58 @@ public static function instance() {
 	 */
 	abstract public function get_logs( $action_id );
 
-	abstract public function init();
 
+	/**
+	 * @codeCoverageIgnore
+	 */
+	public function init() {
+		add_action( 'action_scheduler_stored_action', array( $this, 'log_stored_action' ), 10, 1 );
+		add_action( 'action_scheduler_canceled_action', array( $this, 'log_canceled_action' ), 10, 1 );
+		add_action( 'action_scheduler_before_execute', array( $this, 'log_started_action' ), 10, 1 );
+		add_action( 'action_scheduler_after_execute', array( $this, 'log_completed_action' ), 10, 1 );
+		add_action( 'action_scheduler_failed_execution', array( $this, 'log_failed_action' ), 10, 2 );
+		add_action( 'action_scheduler_failed_action', array( $this, 'log_timed_out_action' ), 10, 2 );
+		add_action( 'action_scheduler_unexpected_shutdown', array( $this, 'log_unexpected_shutdown' ), 10, 2 );
+		add_action( 'action_scheduler_reset_action', array( $this, 'log_reset_action' ), 10, 1 );
+		add_action( 'action_scheduler_execution_ignored', array( $this, 'log_ignored_action' ), 10, 1 );
+	}
+
+	public function log_stored_action( $action_id ) {
+		$this->log( $action_id, __( 'action created', 'action-scheduler' ) );
+	}
+
+	public function log_canceled_action( $action_id ) {
+		$this->log( $action_id, __( 'action canceled', 'action-scheduler' ) );
+	}
+
+	public function log_started_action( $action_id ) {
+		$this->log( $action_id, __( 'action started', 'action-scheduler' ) );
+	}
+
+	public function log_completed_action( $action_id ) {
+		$this->log( $action_id, __( 'action complete', 'action-scheduler' ) );
+	}
+
+	public function log_failed_action( $action_id, \Exception $exception ) {
+		$this->log( $action_id, sprintf( __( 'action failed: %s', 'action-scheduler' ), $exception->getMessage() ) );
+	}
+
+	public function log_timed_out_action( $action_id, $timeout ) {
+		$this->log( $action_id, sprintf( __( 'action timed out after %s seconds', 'action-scheduler' ), $timeout ) );
+	}
+
+	public function log_unexpected_shutdown( $action_id, $error ) {
+		if ( ! empty( $error ) ) {
+			$this->log( $action_id, sprintf( __( 'unexpected shutdown: PHP Fatal error %s in %s on line %s', 'action-scheduler' ), $error[ 'message' ], $error[ 'file' ], $error[ 'line' ] ) );
+		}
+	}
+
+	public function log_reset_action( $action_id ) {
+		$this->log( $action_id, __( 'action reset', 'action_scheduler' ) );
+	}
+
+	public function log_ignored_action( $action_id ) {
+		$this->log( $action_id, __( 'action ignored', 'action-scheduler' ) );
+	}
```

Comment:
❓ I feel like this may be a bit silly, but there's obviously a lot of repetition in most of these methods, especially the ones that don't include any kind of specific error details. Would it be worth implementing a single `__call()` method to handle all of these rather than using separate methods? Perhaps something like this:

```php
	public function __call( $name, $arguments ) {
		$called = preg_match( '#log_(\w+)_action#', $name, $matches );
		if ( ! $called || ! isset( $matches[1] ) ) {
			return;
		}

		$messages = array(
			'stored'    => __( 'action created', 'action-scheduler' ),
			'canceled'  => __( 'action canceled', 'action-scheduler' ),
			'started'   => __( 'action started', 'action-scheduler' ),
			'completed' => __( 'action complete', 'action-scheduler' ),
			'reset'     => __( 'action reset', 'action_scheduler' ),
			'ignored'   => __( 'action ignored', 'action-scheduler' ),
		);

		switch ( $matches[1] ) {
			case 'failed':
				$message = sprintf( __( 'action failed: %s', 'action-scheduler' ), $arguments[1]->getMessage() );
				break;

			case 'timed_out':
				$message = sprintf( __( 'action timed out after %s seconds', 'action-scheduler' ), $arguments[1] );
				break;

			default:
				$message = isset( $messages[ $matches[1] ] ) ? $messages[ $matches[1] ] : "action {$matches[1]}";
				break;
		}

		$this->log( $arguments[0], $message );
	}
```

Just an idea, but certainly not necessary if you'd prefer not to go that route. I just really like avoiding duplication as much as possible 🙂🤷‍♂️

Reply:
> Would it be worth implementing a single __call() method to handle all of these rather than using separate methods?

Maybe, but as with other comments, I'd prefer to keep the code the same when migrating it. If you want to change that, please open a new issue as it's unrelated to the purpose of this PR.

> I just really like avoiding duplication as much as possible

I agree in principle. But I think it's better to be pragmatic about duplication. In this case, the duplication is quite minor and it makes the code far easier to understand than using a `__call()` method.

Right now, I can look at one function and see exactly what it's going to do.

Using a magic method, I first have to figure out why there is no matching method to the one being called, then (hopefully) realise there is a magic method (if I even know what magic methods are, which many people new to PHP won't know about), then parse multiple conditions and data structures in that magic method unrelated to the thing I'm trying to figure out before I can actually see what is being done when calling that method. [_Readability is more important than cleverness or brevity_]9https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#clever-code).